streamz/tests/test_core.py

```diff
@@ -164,6 +164,65 @@ def test_partition():
     assert L == [(0, 1), (2, 3), (4, 5), (6, 7), (8, 9)]
 
 
+def test_partition_timeout():
+    source = Stream()
+    L = source.partition(10, timeout=0.01).sink_to_list()
+
+    for i in range(5):
+        source.emit(i)
+
+    sleep(0.1)
+
+    assert L == [(0, 1, 2, 3, 4)]
+
+
+def test_partition_timeout_cancel():
+    source = Stream()
+    L = source.partition(3, timeout=0.1).sink_to_list()
+
+    for i in range(3):
+        source.emit(i)
+
+    sleep(0.09)
+    source.emit(3)
+    sleep(0.02)
+
+    assert L == [(0, 1, 2)]
+
+    sleep(0.09)
+
+    assert L == [(0, 1, 2), (3,)]
+
+
+def test_partition_key():
+    source = Stream()
+    L = source.partition(2, key=0).sink_to_list()
+
+    for i in range(4):
+        source.emit((i % 2, i))
+
+    assert L == [((0, 0), (0, 2)), ((1, 1), (1, 3))]
+
+
+def test_partition_key_callable():
+    source = Stream()
+    L = source.partition(2, key=lambda x: x % 2).sink_to_list()
+
+    for i in range(10):
+        source.emit(i)
+
+    assert L == [(0, 2), (1, 3), (4, 6), (5, 7)]
+
+
+def test_partition_size_one():
```

Comment:
What is the goal of this test?

Reply:
Here:

https://github.com/roveo/streamz/blob/3747d8b37c909ff53b16f9a3723140698a0dd8a8/streamz/core.py#L1028-L1036

I didn't have `self.n > 1` condition at first and `self._callbacks[key].cancel()` raised KeyError when `n=1`. So it's kind of an edge case test. If you think it's unnecessary, I'll remove it.

Reply:
I'm thinking if we should add a check in `partition` (and update docstring) forcing `n>1`. If `n=1`, the user would be better off avoiding the use of `partition` altogether. What do you think @martindurant?

Reply:
For the sake of test and debug, I think it's OK to allow a redundant partition with n=1.